### ISSUE TITLE
`SyntaxError: Keyword argument repeated` does not get thrown by pegen

### ISSUE THREAD

**Issue by lysnikolaou** [we-like-parsers/cpython#31]
When, for example, parsing `f(1, x=2, *(3, 4), x=5)` the behaviour of the current parse is this:

```
>>> t = 'f(1, x=2, *(3, 4), x=5)'
>>> ast.parse(t)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/lysnikolaou/Repositories/pegen-cpython/Lib/ast.py", line 50, in parse
    return compile(source, filename, mode, flags,
  File "<unknown>", line 1
SyntaxError: keyword argument repeated
```

Pegen currently accepts this.

I don't think this should be a `SyntaxError` at all, but it is and there is a test that tests it as well in `test_grammar.py`.

**Comment by gvanrossum** [comment]
Good call. A simpler example is `f(x=1, x=1)`. I propose to move this check to the bytecode compilation phase.

**Comment by pablogsal** [comment]
> Good call. A simpler example is `f(x=1, x=1)`. I propose to move this check to the bytecode compilation phase.

Will prepare a PR in CPython for that

**Comment by lysnikolaou** [comment]
> > Good call. A simpler example is `f(x=1, x=1)`. I propose to move this check to the bytecode compilation phase.
> 
> Will prepare a PR in CPython for that

Could you please ping me when you do, so that I have a look at it? I'd like to gain an idea of what a change to the bytecode compilation looks like.

**Comment by pablogsal** [comment]
Actually, this is already happening in the compiler stage (particularly in the symbol table):

https://github.com/python/cpython/blob/master/Python/symtable.c#L1028

**Comment by pablogsal** [comment]
I have a PR ready, but the only problem I found is that we cannot point to the incorrect keyword as the keywords in the ast do not remember their col_offset for the keyword itself, only the value.

**Comment by lysnikolaou** [comment]
> I have a PR ready, but the only problem I found is that we cannot point to the incorrect keyword as the keywords in the ast do not remember their col_offset for the keyword itself, only the value.
> 
> This means that the error will look like:
> 
> ```
>   File "lel.py", line 3
>     f(x=1, y=2, z=3, x=1)
>     ^
> SyntaxError: keyword argument repeated
> ```
> 
> instead of
> 
> ```
>  File "lel.py", line 3
>     f(x=1, y=2, z=3, x=1)
>                      ^
> SyntaxError: keyword argument repeated
> ```

I don't think we would be able to exactly point to the error offset in pegen either.

We certainly can't do it in the grammar itself, so we would have to have a helper function that generated the `Call` node for the arguments after checking for duplicate keywords. That means that the whole `Call` node would be parsed first and only then would the error get thrown, thus the error pointer would most probably point to the end of the whole node, right?

**Comment by pablogsal** [comment]
Not necessarily, in the helper we can set a syntax error if we found a duplicate and use our error reporting mechanism immediately if we want, setting accordingly the syntax error.

But I think the correct thing to do is nest the string for the keyword in a `Constant` node therefore providing the location information for the string. 

We would need to fix a couple of places that deal with that like the inspect module. I will open an issue and a PR for discussion in BPO.

**Comment by lysnikolaou** [comment]
> Not necessarily, in the helper we can set a syntax error if we found a duplicate and use our error reporting mechanism immediately if we want, setting accordingly the syntax error.

Oh yes, you're right.

> But I think the correct thing to do is nest the string for the keyword in a Constant node therefore providing the location information for the string.

Whatever you think is best.

**Comment by pablogsal** [comment]
Opened https://bugs.python.org/issue40141 and https://github.com/python/cpython/pull/19283 and

**Comment by pablogsal** [comment]
> Opened https://bugs.python.org/issue40141 and [python#19283](https://github.com/python/cpython/pull/19283) and

Now that this is merged, I opened a PR to move the check to the compiler:

https://github.com/python/cpython/pull/19289/files

**Comment by lysnikolaou** [comment]
Fixed in python#19289.
